### PR TITLE
Better resource scoping rules

### PR DESCRIPTION
**Pull request by expede**
* Account for opposing resources in UCAN validation
* Fix Apache-style logger (much easier to read now)
* Swap local IPFS for remote IPFS only (pain from developing this with staging)
* Drop the `Meta` TH (yay!)

### CODE REVIEW

File: fission-core/library/Fission/Web/Auth/Token/JWT/Proof.hs [hunk 1]
```diff
@@ -37,15 +37,35 @@ signaturesMatch jwt prfJWT =
...
+
+        _ ->
+          if rsc == rscProof
+            then Right jwt
+            else Left ScopeOutOfBounds
+
+    normalizePath path =
+      if "/" `List.isSuffixOf` path
+        then path
+        else path <> "/"
```
Comment (expede) [inline fission-core/library/Fission/Web/Auth/Token/JWT/Proof.hs:68]: Sanitizing filenames vs directories. Specifically handling the case where the proof contains `/foo/bar` but the outer UCAN contains `/foo/barbaz`. This code will normalize to ``/foo/bar/` and `/foo/barbaz/`, which (correctly) does not match

**Comment by expede** [comment]
Just gonna add some tests, but otherwise good to go

File: fission-web-server/library/Fission/Web/Server/Sentry.hs [hunk 2]
```diff
@@ -6,57 +6,63 @@ module Fission.Web.Server.Sentry
...
     sentryRecord time oldRecord =
       oldRecord
-        { srMessage = message time
-        , srLevel   = level
+        { srMessage    = message
+        , srLevel      = level
+        , srTimestamp  = time
+        , srPlatform   = Just "Haskell/Servant"
+        , srServerName = Just $ Servant.baseUrlHost host
+        , srRelease    = Just $ showVersion Fission.version
```
Comment (expede) [inline fission-web-server/library/Fission/Web/Server/Sentry.hs:65]: While I had the hood up, I needed to do some full-system integration debugging (browser, server, &c). So, hey, here's some at least slightly better Sentry logging

File: fission-web-server/library/Fission/Web/Server/Swagger.hs [hunk 3]
```diff
@@ -28,6 +28,8 @@ import           Fission.Web.Server.Internal.Orphanage.BasicAuthData ()
 import           Fission.Web.Server.Internal.Orphanage.HigherOrder   ()
 import           Fission.Web.Server.Internal.Orphanage.RegisterDid   ()
 
+import qualified Paths_fission_web_server                            as Fission
```
Comment (expede) [inline fission-web-server/library/Fission/Web/Server/Swagger.hs:31]: I ran across this the other day — apparently the GHC comes with a bunch of info about the current package available with this weird looking import.

File: fission-web-server/library/Fission/Web/Server/Types.hs [hunk 4]
```diff
@@ -337,14 +337,32 @@ instance IPFS.MonadLocalIPFS Server where
...
+
+          Left err -> do
+            case nonEmpty $ NonEmpty.filter (/= IPFS.URL url) remotes of
+              Nothing -> do
+                logError @Text "🌌🚨 All remote IPFS nodes failed"
+                return $ Left err
+
+              Just fewerRemotes -> do
+                logWarn $ "🌌⚠️  Remote IPFS node " <> hostname <> " failed: " <> textDisplay err
+                go manager fewerRemotes
```
Comment (expede) [inline fission-web-server/library/Fission/Web/Server/Types.hs:365]: Don't just take the first random remote. Fallback to the next if the first fails.

File: fission-web-server/library/Fission/Web/Server/Log.hs [hunk 5]
```diff
@@ -18,9 +18,9 @@ rioApacheLogger ::
   -> Maybe Integer
   -> m ()
 rioApacheLogger Request {..} Status {..} _mayInt =
-  unless (statusCode == 404 && requestHeaderUserAgent == Just "ELB-HealthChecker/2.0") do
+  unless (requestHeaderUserAgent == Just "ELB-HealthChecker/2.0") do
```
Comment (expede) [inline fission-web-server/library/Fission/Web/Server/Log.hs:21]: No more `ELB-HealthChecker/2.0" in the logs 👍

File: fission-web-server/library/Fission/Web/Server/Log.hs [hunk 6]
```diff
@@ -18,9 +18,9 @@ rioApacheLogger ::
...
     if | statusCode >= 500 -> logError formatted
-       | statusCode >= 400 -> logInfo  formatted
+       | statusCode >= 400 -> logWarn  formatted
```
Comment (expede) [inline fission-web-server/library/Fission/Web/Server/Log.hs:23]: Use different pretty colours

File: fission-web-server/library/Fission/Web/Server/Internal/Production.hs [hunk 7]
```diff
@@ -147,10 +147,16 @@ runInProd overrideVerbose action = do
...
+
+  let
+    logOpts =
+      logOptions
+        |> setLogUseLoc  False
+        |> setLogUseTime True
 
   putStrLnIO "   📋 Setting up application logger"
-  withLogFunc (setLogUseTime True logOptions) \baseLogger -> do
+  withLogFunc logOpts \baseLogger -> do
```
Comment (expede) [inline fission-web-server/library/Fission/Web/Server/Internal/Production.hs:159]: Let's make the logs easier to read:

# Before (prod)

Screen Shot 2021-03-05 at 12 08 02 AM

# After (staging)

Screen Shot 2021-03-05 at 12 07 05 AM

File: fission-core/library/Fission/Web/Auth/Token/JWT/Proof/WNFS.hs [hunk 8]
```diff
@@ -0,0 +1,76 @@
+module Fission.Web.Auth.Token.JWT.Proof.WNFS
```
Comment (expede) [inline fission-core/library/Fission/Web/Auth/Token/JWT/Proof/WNFS.hs:1]: The main event

File: fission-web-server/library/Fission/Web/Server/Swagger.hs (hunk 3 above)
Comment (matheus23) [reply fission-web-server/library/Fission/Web/Server/Swagger.hs:31]: Just hopping in here to say: That's a cabal thing: https://www.haskell.org/cabal/release/cabal-2.4.1.0/doc/users-guide/developing-packages.html#accessing-data-files-from-package-code

Comment (expede) [reply fission-web-server/library/Fission/Web/Server/Swagger.hs:31]: Err, Cabal. Yes. Sorry, was typing past midnight 😛

**Comment by expede** [comment]
Leaving this on staging for another day. Investigating what was happening with the staging EU IPFS node this morning, which looked otherwise healthy. Not ruling out the server yet (though unlikely)

File: fission-web-server/library/Fission/Web/Server/IPFS/Client/DAG.hs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    return $ Response cid
+
+dagPut :: MonadRemoteIPFS m => File.Serialized -> m (Either ClientError Response)
+dagPut file = do
+  boundary  <- liftIO genBoundary
+  runRemote (dagClient True (boundary, File.Form "file" file)) >>= \case
+    Left  err -> return $ Left err
+    Right val -> return $ Right val
+  where
+    dagClient = client (Proxy @DAGPut)
```
Comment (expede) [inline fission-web-server/library/Fission/Web/Server/IPFS/Client/DAG.hs:43]: This should move to `ipfs-haskell`

File: fission-web-server/library/Fission/Web/Server/IPFS/Streaming/Pin/Types.hs [hunk 10]
```diff
@@ -1,33 +0,0 @@
...
-  deriving (Eq, Show)
-
-instance Display PinStatus where
-  display status = displayShow status
-
-instance FromJSON PinStatus where
-  parseJSON = withObject "IPFS.PinStatus" \obj -> do
-    pins     <- obj .:? "Pins"     .!= []
-    progress <- obj .:? "Progress"
-    return PinStatus {..}
```
Comment (expede) [inline fission-web-server/library/Fission/Web/Server/IPFS/Streaming/Pin/Types.hs:33]: Moved to `ipfs-haskell`

**Pull request merged**